File: src/c/magma/collect/Deque.c
import magma.Tuple;
import magma.java.JavaOptionals;
import magma.option.Option;
struct Deque<T> {@Deprecated
    default T popOrPanic() {
        return JavaOptionals.from(pop()).map(Tuple::left).orElseThrow();
    }Deque<T> add(T next);boolean isEmpty();Option<Tuple<T, Deque<T>>> pop();Option<T> peek();
}

File: src/java/magma/java/Deque.java
package magma.java;

import magma.Tuple;
import magma.option.Option;

public interface Deque<T> {
    default Deque<T> add(T next) {
        this.list.add(next);
        return this;
    }

    boolean isEmpty();

    Option<Tuple<T, Deque<T>>> pop();

    Option<T> peek();
}
